packages/typegpu/src/data/deepEqual.ts

```diff
@@ -0,0 +1,155 @@
+import type { AnyAttribute } from './attributes.ts';
+import {
+  isDisarray,
+  isLooseData,
+  isLooseDecorated,
+  isUnstruct,
+} from './dataTypes.ts';
+import type { AnyData } from './dataTypes.ts';
+import {
+  isAtomic,
+  isDecorated,
+  isPtr,
+  isWgslArray,
+  isWgslStruct,
+} from './wgslTypes.ts';
+
+/**
+ * Performs a deep comparison of two TypeGPU data schemas.
+ *
+ * @param a The first data schema to compare.
+ * @param b The second data schema to compare.
+ * @returns `true` if the schemas are deeply equal, `false` otherwise.
+ *
+ * @example
+ * ```ts
+ * import { vec3f, struct, deepEqual } from 'typegpu/data';
+ *
+ * const schema1 = struct({ a: vec3f });
+ * const schema2 = struct({ a: vec3f });
+ * const schema3 = struct({ b: vec3f });
+ *
+ * console.log(deepEqual(schema1, schema2)); // true
+ * console.log(deepEqual(schema1, schema3)); // false
+ * ```
+ */
+export function deepEqual(a: AnyData, b: AnyData): boolean {
+  if (a === b) {
+    return true;
+  }
+
+  if (a.type !== b.type) {
+    return false;
+  }
+
+  if (isWgslStruct(a) && isWgslStruct(b)) {
+    const aProps = a.propTypes;
+    const bProps = b.propTypes;
+    const aKeys = Object.keys(aProps);
+    const bKeys = Object.keys(bProps);
+
+    if (aKeys.length !== bKeys.length) {
+      return false;
+    }
+
+    aKeys.sort();
+    bKeys.sort();
+
+    if (aKeys.join() !== bKeys.join()) {
+      return false;
+    }
```

Comment:
The following for loop should already check if all keys in `a` have the same values in `b`. If the set of keys in `a` doesn't match the set of keys in `b`, we'll get `undefined` in some cases, so it should be fine.